components/runtime/src/ts/discord/permissions.ts

```diff
@@ -0,0 +1,206 @@
+    static UseExternalStickers      = 1n << 37n;
+    static SendMessagesInThreads    = 1n << 38n;
+    static UseEmbeddedActivities    = 1n << 39n;
+    static ModerateMembers          = 1n << 40n;
+
+    value: bigint;
+
+    constructor(...data: PermissionResolvable) {
+        this.value = Permissions.resolve(...data);
+    }
+
+    /**
+     * @returns All the Discord permissions as a single bit.
+     */
+    static get all(): bigint {
+        return Flags.reduce((a, b) => a | b[1], 0n);
+    }
+
+    /**
+     * @returns An object containing all Discord permissions.
+     */
+    static entries(): { [key: string]: bigint } {
+        return Flags.reduce<{ [key: string]: bigint }>((a, b) => {
+            a[b[0]] = b[1];
+            return a;
+        }, {});
+    }
+
+    /**
+     * A static method for resolving permissions from strings, numbers and bigints.
+     * @param data The data to resolve permissions from.
+     * @returns The resolved bits.
+     */
+    static resolve(...data: PermissionResolvable): bigint {
+        let result = 0n;
+        const entries = this.entries();
+        for (let i of data) {
+            switch (typeof i) {
+                case 'string': if (i in entries) result |= entries[i]; break;
+                case 'number': result |= BigInt(i); break;
+                default: result |= i; break;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * @param perms The permissions to check for.
+     * @returns True if the current value has any of the given permissions.
+     */
+    hasAny(...perms: PermissionResolvable): boolean {
+        for (let p of perms) {
+            const result = Permissions.resolve(p);
+            if ((this.value & result) === result) return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param perms The permissions to check for.
+     * @returns True if the current value has all of the given permissions.
+     */
+    hasAll(...perms: PermissionResolvable): boolean {
+        for (let p of perms) {
+            const result = Permissions.resolve(p);
+            if ((this.value & result) !== result) return false;
+        }
+        return true;
+    }
+
+    /**
+     * Adds the given permissions to the current value.
+     * @param perms The permissions to add.
+     * @returns The resulting permissions.
+     */
+    add(...perms: PermissionResolvable): bigint {
+        this.value |= Permissions.resolve(...perms);
+        return this.value;
+    }
+
+    /**
+     * Removes the given permissions from the current value.
+     * @param perms The permissions to remove.
+     * @returns The resulting permissions.
+     */
+    remove(...perms: PermissionResolvable): bigint {
+        this.value &= ~Permissions.resolve(...perms);
+        return this.value;
+    }
+
+    /**
+     * @returns An object containing all the permissions the current value has.
+     */
+    serialize(): { [key: string]: bigint } {
```

Comment:
serialize usually has a different meaning (serializing data) so maybe rename it to something like `toMap` since it's the same as `toArray` but with the perm bits? Actually is this function even needed? Is there any case where `toArray` wouldn't be enough?

Reply:
Removed in latest commit. 👍